File: packages/web/src/core/api/homepage.ts
/**
 * Homepage API Client
 * Provides functions to fetch dynamic homepage data from backend
 */

import apiClient from './client';

// ============================================================================
// COLD-START TOLERANCE — public homepage endpoints retry on timeout
// ============================================================================
// Cloud Run scales to zero on staging during low-traffic windows. The first
// hit after idle takes 5-15s to boot (Python imports + DB pool warmup +
// migrations check). With the default axios timeout of 30s, a single in-
// flight request that races the cold start can fail outright. Public
// homepage endpoints are read-only, idempotent, and safe to retry.
const PUBLIC_TIMEOUT_MS = 45_000; // 45s tolerates a typical cold start
const PUBLIC_RETRY_DELAYS = [2_000, 5_000]; // 2s, then 5s — total budget ~52s

function isTimeoutError(error: unknown): boolean {
  if (!error || typeof error !== 'object') return false;
  const e = error as { code?: string; message?: string };
  return e.code === 'ECONNABORTED' || (e.message?.includes('timeout') ?? false);
}

async function withRetryOnTimeout<T>(
  attempt: () => Promise<T>,
  delays: number[] = PUBLIC_RETRY_DELAYS
): Promise<T> {
  let lastError: unknown;
  // First try + retries: delays.length + 1 total attempts
  for (let i = 0; i <= delays.length; i++) {
    try {
      return await attempt();
    } catch (err) {
      lastError = err;
      if (i >= delays.length || !isTimeoutError(err)) {
        throw err; // non-timeout error or budget exhausted
      }
      await new Promise(resolve => setTimeout(resolve, delays[i]));
    }
  }
  throw lastError;
}

/**
 * Homepage statistics response
 */
export interface HomepageStats {
  total_services: number;
  total_ministries: number;
  total_categories: number;
  total_sectors: number;
  last_updated: string;
}

/**
 * Category with service count
 */
export interface CategoryWithServices {
  id: number;
  category_code: string;
  name_es: string;
  name_fr?: string;
  name_en?: string;
  description_es: string | null;
  description_fr?: string | null;
  description_en?: string | null;
  icon: string | null;
  color: string | null;
  service_count: number;
  ministry_name: string | null;
  sector_name: string | null;
}

/**
 * Category directory response
 */
export interface CategoryDirectory {
  total_categories: number;
  total_services: number;
  categories: CategoryWithServices[];
  last_updated: string;
}

/**
 * API Error response
 */
export interface ApiError {
  detail: string;
  status?: number;
}

/**
 * Fetch homepage statistics
 * Returns dynamic counts of services, ministries, categories, and sectors
 *
 * @throws {Error} If the API request fails
 */
export async function getHomepageStats(): Promise<HomepageStats> {
  try {
    const response = await withRetryOnTimeout(() =>
      apiClient.get<HomepageStats>('/homepage/stats', { timeout: PUBLIC_TIMEOUT_MS })
    );
    return response.data;
  } catch (error) {
    console.error('Error fetching homepage stats:', error);

    // Type-safe error handling
    let errorMessage = 'Failed to fetch homepage statistics';
    if (error && typeof error === 'object' && 'response' in error) {
      const axiosError = error as { response?: { data?: { detail?: string } }; message?: string };
      errorMessage = axiosError.response?.data?.detail || axiosError.message || errorMessage;
    } else if (error instanceof Error) {
      errorMessage = error.message;
    }

    throw new Error(errorMessage);
  }
}

/**
 * Fetch category directory with service counts
 * Returns categories sorted by service count (descending)
 *
 * @param language - Language code (es, fr, en). Default: 'es'
 * @throws {Error} If the API request fails
 */
export async function getCategoryDirectory(language: string = 'es'): Promise<CategoryDirectory> {
  try {
    const response = await apiClient.get<CategoryDirectory>('/homepage/categories', {
      params: { language },
    });
    return response.data;
  } catch (error) {
    console.error('Error fetching category directory:', error);

    // Type-safe error handling
    let errorMessage = 'Failed to fetch category directory';
    if (error && typeof error === 'object' && 'response' in error) {
      const axiosError = error as { response?: { data?: { detail?: string } }; message?: string };
      errorMessage = axiosError.response?.data?.detail || axiosError.message || errorMessage;
    } else if (error instanceof Error) {
      errorMessage = error.message;
    }

    throw new Error(errorMessage);
  }
}

/**
 * Get default/fallback stats (used when API fails)
 */
export function getDefaultStats(): HomepageStats {
  return {
    total_services: 0,
    total_ministries: 0,
    total_categories: 0,
    total_sectors: 0,
    last_updated: new Date().toISOString(),
  };
}

/**
 * Get default/fallback category directory (used when API fails)
 */
export function getDefaultCategoryDirectory(): CategoryDirectory {
  return {
    total_categories: 0,
    total_services: 0,
    categories: [],
    last_updated: new Date().toISOString(),
  };
}

/**
 * Service type enum values
 */
export type ServiceType =
  | 'document_processing'
  | 'license_permit'
  | 'residence_permit'
  | 'registration_fee'
  | 'inspection_fee'
  | 'administrative_tax'
  | 'customs_duty'
  | 'declaration_tax';

/**
 * Service details for services-by-type endpoint
 */
export interface ServiceByType {
  id: number;
  service_code: string;
  name_es: string;
  name_fr?: string | null;
  name_en?: string | null;
  tasa_expedicion?: number | null;
}

/**
 * Response for services grouped by type and letter
 */
export interface ServicesByTypeResponse {
  type: string;
  letter?: string | null;
  services: ServiceByType[];
  total: number;
  has_more: boolean;
}

/**
 * Fetch services filtered by service type and optionally by first letter
 *
 * @param type - Service type filter
 * @param options - Optional query parameters
 * @throws {Error} If the API request fails
 */
export async function getServicesByType(
  type: ServiceType,
  options?: {
    letter?: string;
    language?: string;
    limit?: number;
  }
): Promise<ServicesByTypeResponse> {
  try {
    const response = await withRetryOnTimeout(() =>
      apiClient.get<ServicesByTypeResponse>('/homepage/services-by-type', {
        params: {
          type,
          letter: options?.letter,
          language: options?.language || 'es',
          limit: options?.limit || 10,
        },
        timeout: PUBLIC_TIMEOUT_MS,
      })
    );
    return response.data;
  } catch (error) {
    console.error('Error fetching services by type:', error);

    // Type-safe error handling
    let errorMessage = 'Failed to fetch services by type';
    if (error && typeof error === 'object' && 'response' in error) {
      const axiosError = error as { response?: { data?: { detail?: string } }; message?: string };
      errorMessage = axiosError.response?.data?.detail || axiosError.message || errorMessage;
    } else if (error instanceof Error) {
      errorMessage = error.message;
    }

    throw new Error(errorMessage);
  }
}

/**
 * Get default/fallback services by type response (used when API fails)
 */
export function getDefaultServicesByType(type: ServiceType): ServicesByTypeResponse {
  return {
    type,
    letter: null,
    services: [],
    total: 0,
    has_more: false,
  };
}

// ============================================================================
// MINISTRY TYPES AND FUNCTIONS
// ============================================================================

/**
 * Ministry item with service counts
 */
export interface MinistryItem {
  id: number;
  ministry_code: string;
  name: string;
  description: string | null;
  icon: string | null;
  color: string | null;
  is_active: boolean;
  service_count: number;
  sector_count: number;
  category_count: number;
}

/**
 * Ministry directory response
 */
export interface MinistryDirectory {
  total_ministries: number;
  total_services: number;
  ministries: MinistryItem[];
}

/**
 * Ministry service item
 */
export interface MinistryServiceItem {
  id: number;
  service_code: string;
  name: string;
  description: string | null;
  expedition_price: number;
  renewal_price: number;
  category_name: string | null;
  sector_name: string | null;
  service_type: string;
}

/**
 * Ministry details response with paginated services
 */
export interface MinistryDetails {
  id: number;
  ministry_code: string;
  name: string;
  description: string | null;
  icon: string | null;
  color: string | null;
  is_active: boolean;
  service_count: number;
  sector_count: number;
  category_count: number;
  services: MinistryServiceItem[];
  total_pages: number;
  current_page: number;
}

/**
 * Fetch ministry directory with service counts
 *
 * @param language - Language code (es, fr, en). Default: 'es'
 * @throws {Error} If the API request fails
 */
export async function getMinistryDirectory(language: string = 'es'): Promise<MinistryDirectory> {
  try {
    const response = await apiClient.get<MinistryDirectory>('/homepage/ministries', {
      params: { language },
    });
    return response.data;
  } catch (error) {
    console.error('Error fetching ministry directory:', error);

    let errorMessage = 'Failed to fetch ministry directory';
    if (error && typeof error === 'object' && 'response' in error) {
      const axiosError = error as { response?: { data?: { detail?: string } }; message?: string };
      errorMessage = axiosError.response?.data?.detail || axiosError.message || errorMessage;
    } else if (error instanceof Error) {
      errorMessage = error.message;
    }

    throw new Error(errorMessage);
  }
}

/**
 * Fetch ministry details with paginated services
 *
 * @param ministryId - Ministry ID
 * @param options - Optional query parameters
 * @throws {Error} If the API request fails
 */
export async function getMinistryDetails(
  ministryId: number,
  options?: {
    language?: string;
    page?: number;
    limit?: number;
  }
): Promise<MinistryDetails> {
  try {
    const response = await apiClient.get<MinistryDetails>(`/homepage/ministry/${ministryId}`, {
      params: {
        language: options?.language || 'es',
        page: options?.page || 1,
        limit: options?.limit || 12,
      },
    });
    return response.data;
  } catch (error) {
    console.error('Error fetching ministry details:', error);

    let errorMessage = 'Failed to fetch ministry details';
    if (error && typeof error === 'object' && 'response' in error) {
      const axiosError = error as { response?: { data?: { detail?: string } }; message?: string };
      errorMessage = axiosError.response?.data?.detail || axiosError.message || errorMessage;
    } else if (error instanceof Error) {
      errorMessage = error.message;
    }

    throw new Error(errorMessage);
  }
}

/**
 * Get default/fallback ministry directory (used when API fails)
 */
export function getDefaultMinistryDirectory(): MinistryDirectory {
  return {
    total_ministries: 0,
    total_services: 0,
    ministries: [],
  };
}

File: packages/backend/app/modules/homepage/models/homepage.py
"""
Homepage Module - Pydantic Models
Defines request/response models for homepage statistics and category directory
"""

from pydantic import BaseModel, Field
from typing import List, Optional


class HomepageStats(BaseModel):
    """Dynamic homepage statistics calculated from database"""
    total_services: int = Field(..., description="Total number of active fiscal services")
    total_ministries: int = Field(..., description="Total number of active ministries")
    total_categories: int = Field(..., description="Total number of active categories")
    total_sectors: int = Field(..., description="Total number of active sectors")
    last_updated: str = Field(..., description="ISO timestamp of when stats were calculated")


class CategoryWithServices(BaseModel):
    """Category with service count and translations"""
    id: int = Field(..., description="Category ID")
    category_code: str = Field(..., description="Category code")
    name: str = Field(..., description="Category name (translated)")
    description: Optional[str] = Field(None, description="Category description (translated)")
    icon: Optional[str] = Field(None, description="Category icon")
    color: Optional[str] = Field(None, description="Category color")
    service_count: int = Field(..., description="Number of active services in this category")
    ministry_name: Optional[str] = Field(None, description="Associated ministry name")
    sector_name: Optional[str] = Field(None, description="Associated sector name")


class CategoryDirectory(BaseModel):
    """Category directory response"""
    total_categories: int = Field(..., description="Total number of categories")
    total_services: int = Field(..., description="Total number of services across all categories")
    categories: List[CategoryWithServices] = Field(..., description="List of categories with service counts")
    last_updated: str = Field(..., description="ISO timestamp")


class ServiceByType(BaseModel):
    """Service details for services-by-type endpoint"""
    id: int = Field(..., description="Service ID")
    service_code: str = Field(..., description="Service code")
    name_es: str = Field(..., description="Service name (Spanish)")
    name_fr: Optional[str] = Field(None, description="Service name (French)")
    name_en: Optional[str] = Field(None, description="Service name (English)")
    tasa_expedicion: Optional[float] = Field(None, description="Expedition fee in GNF")


class ServicesByTypeResponse(BaseModel):
    """Response for services grouped by type and letter"""
    type: str = Field(..., description="Service type filter")
    letter: Optional[str] = Field(None, description="First letter filter (A-Z)")
    services: List[ServiceByType] = Field(..., description="List of services")
    total: int = Field(..., description="Total services for this type/letter")
    has_more: bool = Field(..., description="Whether there are more services beyond limit")


# ============================================================================
# MINISTRY MODELS
# ============================================================================

class MinistryItem(BaseModel):
    """Ministry item with service count"""
    id: int = Field(..., description="Ministry ID")
    ministry_code: str = Field(..., description="Ministry code")
    name: str = Field(..., description="Ministry name (translated)")
    description: Optional[str] = Field(None, description="Ministry description (translated)")
    icon: Optional[str] = Field(None, description="Ministry icon")
    color: Optional[str] = Field(None, description="Ministry color")
    is_active: bool = Field(True, description="Whether ministry is active")
    service_count: int = Field(0, description="Number of services under this ministry")
    sector_count: int = Field(0, description="Number of sectors under this ministry")
    category_count: int = Field(0, description="Number of categories under this ministry")


class MinistryDirectory(BaseModel):
    """Ministry directory response"""
    total_ministries: int = Field(..., description="Total number of ministries")
    total_services: int = Field(..., description="Total number of services across all ministries")
    ministries: List[MinistryItem] = Field(..., description="List of ministries with service counts")


class MinistryServiceItem(BaseModel):
    """Service item for ministry details"""
    id: int = Field(..., description="Service ID")
    service_code: str = Field(..., description="Service code")
    name: str = Field(..., description="Service name (translated)")
    description: Optional[str] = Field(None, description="Service description (translated)")
    expedition_price: float = Field(0, description="Expedition price")
    renewal_price: float = Field(0, description="Renewal price")
    category_name: Optional[str] = Field(None, description="Category name")
    sector_name: Optional[str] = Field(None, description="Sector name")
    service_type: str = Field(..., description="Service type")


class MinistryDetails(BaseModel):
    """Complete ministry details with services"""
    id: int = Field(..., description="Ministry ID")
    ministry_code: str = Field(..., description="Ministry code")
    name: str = Field(..., description="Ministry name (translated)")
    description: Optional[str] = Field(None, description="Ministry description (translated)")
    icon: Optional[str] = Field(None, description="Ministry icon")
    color: Optional[str] = Field(None, description="Ministry color")
    is_active: bool = Field(True, description="Whether ministry is active")
    service_count: int = Field(0, description="Total services")
    sector_count: int = Field(0, description="Total sectors")
    category_count: int = Field(0, description="Total categories")
    services: List[MinistryServiceItem] = Field([], description="Services list (paginated)")
    total_pages: int = Field(1, description="Total pages")
    current_page: int = Field(1, description="Current page")


# ============================================================================
# SEARCH MODELS
# ============================================================================

class SearchRequest(BaseModel):
    """Search request payload"""
    q: Optional[str] = Field(None, description="Search query")
    category_id: Optional[int] = Field(None, description="Filter by category ID")
    category_code: Optional[str] = Field(None, description="Filter by category code")
    ministry_id: Optional[int] = Field(None, description="Filter by ministry ID")
    service_type: Optional[str] = Field(None, description="Filter by service type")
    min_price: Optional[float] = Field(None, description="Minimum price filter")
    max_price: Optional[float] = Field(None, description="Maximum price filter")
    calculation_methods: Optional[List[str]] = Field(None, description="Filter by calculation methods (e.g., ['percentage_based', 'formula_based'])")
    sort_by: str = Field("relevance", description="Sort by: relevance, name, price")
    sort_order: str = Field("asc", description="Sort order: asc, desc")
    page: int = Field(1, ge=1, description="Page number")
    limit: int = Field(20, ge=1, le=100, description="Results per page")
    include_facets: bool = Field(True, description="Include facets in response")
    language: str = Field("es", description="Language code")


class SearchResultItem(BaseModel):
    """Single search result"""
    id: int
    name: str
    description: Optional[str] = None
    category_name: str
    ministry_name: Optional[str] = None
    sector_name: Optional[str] = None
    service_type: str
    expedition_price: float = 0
    renewal_price: float = 0
    processing_time_days: int = 30
    status: str = "active"
    calculation_method: str = "fixed_expedition"


class FacetItem(BaseModel):
    """Facet item for filtering"""
    id: Optional[int] = None
    code: Optional[str] = None
    name: Optional[str] = None
    type: Optional[str] = None
    count: int = 0


class SearchFacets(BaseModel):
    """Search facets"""
    categories: List[FacetItem] = []
    ministries: List[FacetItem] = []
    service_types: List[FacetItem] = []
    price_ranges: List[FacetItem] = []


class BundleResultItem(BaseModel):
    """Bundle search result (commercial license package)"""
    id: str
    name: str
    description: Optional[str] = None
    bundle_code: str
    commerce_type: str
    item_count: int = 0


class SearchResponse(BaseModel):
    """Search response"""
    success: bool = True
    query: str = ""
    total_results: int = 0
    page: int = 1
    limit: int = 20
    total_pages: int = 0
    results: List[SearchResultItem] = []
    bundles: List[BundleResultItem] = []
    facets: Optional[SearchFacets] = None
    suggestions: List[str] = []
    execution_time_ms: float = 0
    cached: bool = False
